mansrc/consensusFinder.1
'\" t
.\"     Title: consensusfinder
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5.dev
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "CONSENSUSFINDER" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
consensusfinder \- consensus finder
.SH "SYNOPSIS"
.sp
\fBconsensusFinder\fP input_fasta_file [options]
.SH "OPTIONS"
.sp
\fB\-l\fP
.RS 4
pattern length
.RE
.sp
\fB\-p\fP
.RS 4
p\-value for finding the significant strings
.RE
.sp
\fB\-d\fP
.RS 4
delta to find the relevant strings
.RE
.sp
\fB\-s\fP
.RS 4
starting position
.RE
.sp
\fB\-e\fP
.RS 4
ending position
.RE
.sp
\fB\-m\fP
.RS 4
no. of mismatches to consider while finding neighbours
.RE
.sp
\fB\-c\fP
.RS 4
string to consider for analysis
.RE
.sp
\fB\-n\fP
.RS 4
specify the number of consensus patterns to store
.RE
.sp
\fB\-t\fP
.RS 4
print the histogram
.RE
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.
.SH "ADDITIONAL DOCUMENTATION"
.sp
An exhaustive documentation can be found in the file /usr/share/augustus/README.TXT.

HISTORY.TXT
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic licence only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently
     - hints can be loaded to the sqlite database
     - update of bam2wig to latest samtools
List of changes from version 3.0.2 to 3.0.3 (until August 28th, 2014)
     - made dependency on boost libraries an option
     - softmasking support (--softmasking=1 improves speed by almost a factor of 2 and exon level sensitivity by 0.01 compared to hard masking on human)
     - comparative gene finding with UTRs
     - new sqlite database option for genome input in comparative mode besides flat files and MySQL
     - new species: zebrafish (Danio rerio), camponotus_floridanus (an ant) contributed by Shishir K Gupta, s_aureus (Staphylococcus aureus)
     - updated parameters: thermoanaerobacter_tengcongensis, E_coli_K12
     - bugfix in bacterial model: length probability of very long genes was wrong
     - other bugfixes in comparative mode
     - removal of mysql++ code from package: this is available as Debian/Ubuntu package
List of changes from version 3.0 to 3.0.2 (until February 22nd, 2014)
     - inclusion of scanner and parser code for Newick tree format (GPL licence), easier installation of 
       comparative augustus functionality
     - therefore release under more restrictive GPL licence (the plan is to replace this later
       and switch back to the more permissive and non-viral artistic open source licence)
     - drop requirement to set environment variable AUGUSTUS_CONFIG_PATH
     - bugfixes (etraining: "not Genbank format", gtf2gff.pl, compilation warning)
List of changes from version 2.7 to 3.0 (until January 8th, 2014)
     - comparative gene prediction (cgp) mode to predict genes in several aligned genomes simultaneously (see README-cgp.txt)
     - load2db program to load genomes and hint GFF files into a database for random access retrieval (e.g. with new program getSeq)
     - new parameter --temperature for heating the distribution and obtaining more gene structures during sampling
     - bacterial genes may now also be nested
     - new script weedMaf.pl that throws out gap-only columns from MSAs
     - added prokaryotic template parameters for training new prokaryotic species (on the basis of E_coli_K12 parameters)
     - new species coyote_tobacco (Nicotiana attenuata), thermoanaerobacter_tengcongensis, wheat
     - ATG is not anymore the only start codon
     - filterMaf.pl takes an alignment in MAF format and extracts alignment blocks given a subset of species or a genomic interval
     - bugfixes (in particular segfault on MAC, small memory leaks)

src/consensusFinder.cc
#include "consensus.hh"

#include <getopt.h>
#include <cstdlib>

using namespace std;

void print_consensus(vector<histogram_data> final_list){
  int i,j;
  cout <<" Printing the consensus patterns, f values, expected values and neighbours" << endl;
  for(i=0;i<final_list.size();i++){
    cout<<final_list[i].consensus_pattern<<'\t'<<final_list[i].f_value<< '\t'<< final_list[i].mean_value<< endl;
    cout<< "The mismatch neighbours are:" << endl;
    for(j=0;j<final_list[i].neighbours.size();j++)
      cout<<final_list[i].neighbours[j]<<'\t'<<final_list[i].freq_neighbours[j]<<'\t'<<final_list[i].mean_neighbours[j]<<endl;
    cout <<endl<<endl;
  }
}

void usage(){
  cout<<"Usage:\n ./consensusFinder input fasta file [options] \n -l:pattern length\n -p:p value for finding the significant strings \n -d:delta to find the relevant strings\n -s:starting position\n -e:ending position\n -m:no. of mismatches to consider while finding neighbours\n -c:string to consider for analysis\n -n:to specify the number of consensus patterns to store\n -t:to print the histogram\n -h:help\n"; 

}
int main(int argc, char** argv){
  vector<histogram_data> final_list;
  string filename;
  int pattern_size=6,no_iterations=10;
  float p_value=.01;
  float delta=2;
  int max_allowed_mismatch=1,start=0,end=0,plot_hist=0;
  string pattern="none";
  int c,index,longIndex;
  static const struct option longOpts[]={
    {"pattern_size",required_argument,NULL,'l'},
    {"p_value",required_argument,NULL,'p'},
    {"start",required_argument,NULL,'s'},
    {"end",required_argument,NULL,'e'},
    {"delta",required_argument,NULL,'d'},
    {"mismatch",required_argument,NULL,'m'},
    {"no_consensus",required_argument,NULL,'n'},
    {"print_hist",no_argument,NULL,'t'},
    {"string_to_analyse",required_argument,NULL,'c'},
    {"help",no_argument,NULL,'h'},
    { NULL, no_argument, NULL, 0 }
    };
  opterr = 0;  
  while ((c = getopt_long (argc, argv, ":thl:p:d:m:s:n:e:c:",longOpts, &longIndex)) != -1){
    switch (c)
      {
      case 'l':
	pattern_size = atoi(optarg);
	break;
      case 'p':
	p_value = atof(optarg);
	break;
      case 's':
	start = atoi(optarg);
	break;
      case 'e':
	end = atoi(optarg);
	break;
      case 'd':
	delta = atof(optarg);
	break;
      case 'm':
	max_allowed_mismatch = atoi(optarg);
	break;
      case 'n':
	no_iterations = atoi(optarg);
	break;
      case 't':
	plot_hist=1;
	break;
      case 'c':
	pattern=optarg;
	break;
      case 'h':    
	usage();
	exit(0);
	break;
      case 0:     /* long option without a short arg */
	cout<<" Wrong option"<<endl;
	usage();
	break;

      default:
	cout<<" Wrong option"<<endl;
	usage();
	exit(0);
      } 
  }  
  if((argc-optind)==0){
    usage();
    exit(0);
  }
  for (index = optind; index < argc; index++)
    filename=argv[index];

  consensus consensus1(start,end);
  consensus1.set_file_name(filename);

  if(pattern=="none"){
    consensus1.set_values(pattern_size,p_value,delta,max_allowed_mismatch,no_iterations);    
    consensus1.start();
  }
  else
    consensus1.analyse(pattern,p_value,delta,max_allowed_mismatch);
  final_list=consensus1.get_consensus();
  if(plot_hist==1)
    consensus1.plot_histogram();
  print_consensus(final_list);
  return 0;
}


